File: src/test/java/com/conference/website/dto/TestDtoConversions.java
package com.conference.website.dto;

import com.conference.website.domain.*;

import java.util.List;

public final class TestDtoConversions {

    public static SpeakerDto toDto(Long id, CreateSpeakerRequest request) {
        return new SpeakerDto(id, request.name(), request.email(), request.company(), request.bio());
    }

    public static SpeakerDto toDto(Speaker speaker) {
        return new SpeakerDto(speaker.getId(), speaker.getName(), speaker.getEmail(), speaker.getCompany(), speaker.getBio());
    }


    public static TagDto toDto(Tag tag) {
        return new TagDto(tag.getId(), tag.getName());
    }

    public static RatingDto toDto(Rating rating) {
        return new RatingDto(rating.getId(), rating.getReviewerName(), rating.getScore(), rating.getComment());
    }


    public static ScheduleSlotDto toDto(Long id, ScheduleSlotRequest request) {
        return new ScheduleSlotDto(id, request.roomName(), request.startTime(), request.endTime());
    }

    public static ScheduleSlotDto toDto(ScheduleSlot slot) {
        return new ScheduleSlotDto(slot.getId(), slot.getRoomName(), slot.getStartTime(), slot.getEndTime());
    }


    public static TalkDto toDto(Talk talk) {
        return new TalkDto(talk.getId(), talk.getTitle(), talk.getAbstractText(), talk.getLevel(), talk.getDurationMinutes(), toDto(talk.getPrimarySpeaker()), talk.getCoSpeakers().stream().map(TestDtoConversions::toDto).toList(), talk.getTags().stream().map(TestDtoConversions::toDto).toList(), talk.getRatings().stream().map(TestDtoConversions::toDto).toList(), toDto(talk.getScheduleSlot()), talk.getRatings().stream().mapToInt(i-> i.getScore()).average().getAsDouble() , Integer.valueOf(talk.getRatings().size()).longValue());
    }


    public static TalkDto toDto(Long id, SpeakerDto speaker, List<SpeakerDto>   coSpeakers,
                                List<TagDto> tagIds, List<RatingDto> ratings, ScheduleSlotDto scheduleSlotDto, CreateTalkRequest request) {
        return new TalkDto(id, request.title(), request.abstractText(), request.level(), request.durationMinutes(), speaker, coSpeakers, tagIds, ratings,scheduleSlotDto, ratings.stream().mapToInt(i-> i.score()).average().getAsDouble(),  ratings.stream().count());
    }

    public static TalkDto toDto(Long id, SpeakerDto speaker, ScheduleSlotDto scheduleSlotDto, CreateTalkRequest request) {
        return new TalkDto(id, request.title(), request.abstractText(), request.level(), request.durationMinutes(), speaker, List.of(), List.of(), List.of(),scheduleSlotDto, 0.0, 0L);
    }


    public static TalkDto toDto(Long id, CreateTalkRequest request) {
        return new TalkDto(id, request.title(), request.abstractText(), request.level(), request.durationMinutes(), request.primarySpeaker(), request.coSpeakers(), request.tags(), List.of(), null, 0.0, 0L);
    }
}

File: src/test/kotlin/com/conference/website/dto/TestDtoConversions.kt
package com.conference.website.dto

import com.conference.website.domain.Rating
import com.conference.website.domain.ScheduleSlot
import com.conference.website.domain.Speaker
import com.conference.website.domain.Tag

/**
 * The inferior conversion helper: one overload per source type, all crammed into a
 * single static-style holder. Kotlin would rather have extension functions on the
 * types themselves (see `dto/DtoConversions.kt` in main) — that is the point of the
 * contrast.
 */
object TestDtoConversions {

    fun toDto(id: Long?, request: CreateSpeakerRequest) =
        SpeakerDto(id, request.name, request.email, request.company, request.bio)

    fun toDto(speaker: Speaker) =
        SpeakerDto(speaker.id, speaker.name, speaker.email, speaker.company, speaker.bio)

    fun toDto(tag: Tag) = TagDto(tag.id, tag.name)

    fun toDto(rating: Rating) =
        RatingDto(rating.id, rating.reviewerName, rating.score, rating.comment)

    fun toDto(id: Long?, request: ScheduleSlotRequest) =
        ScheduleSlotDto(id, request.roomName, request.startTime, request.endTime)

    fun toDto(slot: ScheduleSlot) =
        ScheduleSlotDto(slot.id, slot.roomName, slot.startTime, slot.endTime)

    fun toDto(id: Long?, request: CreateTalkRequest) =
        TalkDto(
            id,
            request.title,
            request.abstractText,
            request.level,
            request.durationMinutes,
            request.primarySpeaker,
            request.coSpeakers,
            request.tags,
            emptyList(),
            null,
            0.0,
            0L
        )
}
